File: cs50/pset6/crack/crack.py
# A program to crack simple passwords that use DES-based crypt.

import crypt
import sys
import os


def main():
    """
    Takes a command line argument, an encrypted password (aka shadow).

    Assumes shadow came from crypt function with DES-based encryption.

    First, checks shadow against the hashes of popular passwords and
    common English words. You can easily add more documents to the doc list
    of words to check. Each doc must be a text file with one word per line.

    Second, tries brute force on the password, using every combination of
    five letters [a-zA-Z].

    *Five letters takes about 45 to 55 minutes to fully crack using brute force.
    That's 52^5 calculations, or north of 380 million potential passwords to
    check
    """

    # More or less than one command line argument returns error & quits
    if len(sys.argv) != 2:
        print("Enter one string as shadow")
        return False

    # shadow is the hashed password; in crypt, salt is first two chars of shadow
    shadow = sys.argv[1]
    salt = shadow[0] + shadow[1]

    # List of text files to check before brute forcing
    doc_list = [
        "test_words.txt",
        "common_passwords.txt",
        "five_letters.txt"
    ]

    # Check each line in each doc using fcheck function (below)
    for doc in doc_list:
        print(f"Now checking {doc}...")
        check = fcheck(shadow, salt, doc)
        if check is not None:
            print(f"{check}")
            return 0

    # Failing fcheck, use brute force
    password = cracker(shadow, salt)
    print(f"{password}")


def fcheck(shadow, salt, fname):
    """
    Takes an shadow, salt, and the name of a file.
    Goes through file line-by-line, checking potential passwords.
    Returns password as a string, if found.
    """
    if os.stat(fname).st_size > 0:
        file = open(fname, "r")
        for line in file:
            if (line.isalpha()):
                if crypt.crypt(line, salt) == shadow:
                    return line
                elif crypt.crypt(line.upper(), salt) == shadow:
                    return line.upper()
        file.close()
    return None


def cracker(shadow, salt):
    """
    Takes shadow and salt.
    Iterates through alphabet [a-zA-Z], first with one-character passwords.
    Tries all combinations of two-character passwords.
    Tries all combinations of three-character passwords.
    Continues until it has exhausted all five character passwords.
    Returns password as a string, if found.
    """
    alphabet = "etaoinsrhldcuETAOINSRHLDCUmfpgwybvkxjqzMFPGWYBVKXJQZ"

    # Crack a one-character password [a-zA-Z]:
    for i in alphabet:
        if crypt.crypt(i, salt) == shadow:
            return i

    # Two-character passwords
    for i in alphabet:
        for j in alphabet:
            pw = i + j
            if crypt.crypt(pw, salt) == shadow:
                return pw

    # Three-character passwords:
    for i in alphabet:
        for j in alphabet:
            for k in alphabet:
                pw = i + j + k
                if crypt.crypt(pw, salt) == shadow:
                    return pw

    # Four-character passwords:
    for i in alphabet:
        for j in alphabet:
            for k in alphabet:
                for l in alphabet:
                    pw = i + j + k + l
                    if crypt.crypt(pw, salt) == shadow:
                        return pw

    # Five-character passwords:
    for i in alphabet:
        for j in alphabet:
            for k in alphabet:
                for l in alphabet:
                    for m in alphabet:
                        pw = i + j + k + l + m
                        if crypt.crypt(pw, salt) == shadow:
                            return pw

    return "No password found"


if __name__ == '__main__':
    main()

File: cs50/pset2/crack/crack.c
//A program to crack simple passwords that use crypt.
//I know there has to be some type of recursive function to make this work better, but I'm happy just to have a working program.

#define _XOPEN_SOURCE

#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

int checkCrypt(string hash, string shadow);

int main(int argc, string argv[])
{
    //More or less than one argument yields error and quits program.
    if (argc != 2)
    {
        printf("Enter one string as the encrypted password.\n");
        return 1;
    }

    //Salt is first two digits of argv[1]
    char salt[3];
    salt[0] = (char) argv[1][0];
    salt[1] = (char) argv[1][1];
    salt[2] = '\0';

    char key[6];
    string alphabet = "etaoinsrhldcuETAOINSRHLDCUmfpgwybvkxjqzMFPGWYBVKXJQZ";

    //Crack a password of a single alpha character [a-zA-Z]
    for (int i = 0; i < 52; i++)
    {
        key[0] = alphabet[i];
        key[1] = '\0';

        if ((checkCrypt(crypt(key, salt), argv[1])) == 0)
        {
            goto end_nested_loop;
        }
    }

    //Two character passwords:
    for (int i = 0; i < 52; i++)
    {
        key[0] = alphabet[i];

        for (int j = 0; j < 52; j++)
        {

            key[1] = alphabet[j];
            key[2] = '\0';

            if ((checkCrypt(crypt(key, salt), argv[1])) == 0)
            {
                goto end_nested_loop;
            }
        }
    }

    //Three character passwords:
    for (int i = 0; i < 52; i++)
    {
        key[0] = alphabet[i];

        for (int j = 0; j < 52; j++)
        {
            key[1] = alphabet[j];

            for (int k = 0; k < 52; k++)
            {
                key[2] = alphabet[k];
                key[3] = '\0';

                if ((checkCrypt(crypt(key, salt), argv[1])) == 0)
                {
                    goto end_nested_loop;
                }
            }
        }
    }

    //Four character passwords:
    for (int i = 0; i < 52; i++)
    {
        key[0] = alphabet[i];

        for (int j = 0; j < 52; j++)
        {
            key[1] = alphabet[j];

            for (int k = 0; k < 52; k++)
            {
                key[2] = alphabet[k];
                for (int l = 0; l < 52; l++)
                {
                    key[3] = alphabet[l];
                    key[4] = '\0';

                    if ((checkCrypt(crypt(key, salt), argv[1])) == 0)
                    {
                        goto end_nested_loop;
                    }
                }
            }
        }
    }

    //Five character passwords:
    for (int i = 0; i < 52; i++)
    {
        key[0] = alphabet[i];

        for (int j = 0; j < 52; j++)
        {
            key[1] = alphabet[j];

            for (int k = 0; k < 52; k++)
            {
                key[2] = alphabet[k];
                for (int l = 0; l < 52; l++)
                {
                    key[3] = alphabet[l];

                    for (int m = 0; m < 52; m++)
                    {
                        key[4] = alphabet[m];
                        key[5] = '\0';

                        if ((checkCrypt(crypt(key, salt), argv[1])) == 0)
                        {
                            goto end_nested_loop;
                        }
                    }
                }
            }
        }
    }

end_nested_loop:
    printf("%s\n", key);
}

//Check if hash matches input, character by character
int checkCrypt(string hash, string shadow)
{
    int x = 0;

    do
    {
        if (hash[x] == shadow[x])
        {
            x++;
        }
        else
        {
            return 1;
        }
    }
    while (x < strlen(shadow));

    return 0;
}